core/core-format.el

```diff
@@ -0,0 +1,146 @@
+      (compilation-mode)
+      (display-buffer errbuf))))
+
+(defun spacemacs-format//kill-error-buffer (errbuf)
+  "Kill buffer ERRBUF."
+  (let ((win (get-buffer-window errbuf)))
+    (if win
+        (quit-window t win)
+      (with-current-buffer errbuf
+        (erase-buffer))
+      (kill-buffer errbuf))))
+
+(defun spacemacs-format//format-buffer (name program args)
+  (let* ((ext (file-name-extension buffer-file-name t))
+         (bufferfile (make-temp-file name nil ext))
+         (outputfile (make-temp-file name nil ext))
+         (errorfile (make-temp-file name nil ext))
+         (errbuf (when dotspacemacs-format-show-errors
+                   (get-buffer-create (format "*%s errors*" name))))
+         (patchbuf (get-buffer-create (format "*%s patch*" name)))
+         (coding-system-for-read 'utf-8)
+         (coding-system-for-write 'utf-8))
+    (unwind-protect
+        (save-restriction
+          (widen)
+          (write-region nil nil bufferfile)
+          (when errbuf
+            (with-current-buffer errbuf
+              (setq buffer-read-only nil)
+              (erase-buffer)))
+          (with-current-buffer patchbuf
+            (erase-buffer))
+          (if (zerop (apply 'call-process program bufferfile
+                            (list (list :file outputfile) errorfile)
+                            nil args))
+              (progn
+                (call-process-region (point-min) (point-max)
+                                     "diff" nil patchbuf nil
+                                     "-n" "--strip-trailing-cr" "-" outputfile)
+                (spacemacs-format//format-apply-rcs-patch patchbuf)
+                (message "Applied %s" name)
+                (when errbuf (spacemacs-format//kill-error-buffer errbuf)))
+            (message "Could not apply %s" name)
+            (when errbuf
+              (spacemacs-format//format-process-errors errorfile errbuf))))
+      (kill-buffer patchbuf)
+      (delete-file errorfile)
+      (delete-file bufferfile)
+      (delete-file outputfile))))
+
+(defmacro spacemacs|define-fmt-tool (&rest props)
+  "Generate an interactive format command `spacemacs/xxx-format' where xxx is ID.
+
+Available PROPS:
+
+`:id ID'
+A symbol naming the formatter function. Must be started with the layer name.
+
+`:name NAME'
+The formatter name.
+
+`:program PROGRAM'
+A literal string which holds the program path.
+
+`:args ARGS'
+A list of arguments or a function which evaluates to a list of arguments."
+  (let* ((id      (spacemacs/mplist-get-value props :id))
+         (name    (spacemacs/mplist-get-value props :name))
+         (program (spacemacs/mplist-get-value props :program))
+         (args    (spacemacs/mplist-get-value props :args))
+         (format-func-name (intern (format "spacemacs/%S-format" id))))
+    ;; define format function
+    `(defun ,format-func-name ()
+       (interactive)
+       (let* (,(if (functionp args) `(args (,args)) `(args ',args)))
```

Comment:
I think your macro contains a memory anomaly, this is the variable `args` should be an uninterned symbol to avoid that a possible local variable `args` in the formating function is overwritten by the one in your macro.

See https://www.gnu.org/software/emacs/manual/html_node/elisp/Surprising-Local-Vars.html#Surprising-Local-Vars for a detailed explanation of the problem and how to resolve it.